bottlerocket/agents/src/userdata.rs

```diff
@@ -0,0 +1,106 @@
+use resource_agent::provider::{IntoProviderError, ProviderResult, Resources};
+use toml::{map::Entry, Value};
+
+/// This modifies the first given toml Value by inserting any values from the second Value.
+///
+/// This is done recursively.  Any time a scalar or array is seen, the left side is set to match
+/// the right side.  Any time a table is seen, we iterate through the keys of the tables; if the
+/// left side does not have the key from the right side, it's inserted, otherwise we recursively
+/// merge the values in each table for that key.
+///
+/// If at any point in the recursion the data types of the two values does not match, we error.
+pub fn merge_values<'a>(merge_into: &'a mut Value, merge_from: &'a Value) -> ProviderResult<()> {
```

Comment:
I prefer left-to-right, `from, to` ordering (I always mess up when creating symlinks `ln -s to from`)